### DISCUSSION TITLE
Completion blocks in play() methods not called

### DISCUSSION THREAD

**Discussion by oliveira-marcio** [airbnb/lottie-ios#1997] [General]
I am wondering if the completion blocks inside some `play()` methods from the animation view might never be called because some of the guard lets that just return without calling the completion, therefore I am assuming that it might lead to the animation being stuck, especially when chaining multiple `play()` calls.

Example in `LottieAnimationView.swift`:
```
open func play(
    fromProgress: AnimationProgressTime? = nil,
    toProgress: AnimationProgressTime,
    loopMode: LottieLoopMode? = nil,
    completion: LottieCompletionBlock? = nil)
  {
    guard let animation = animation else {
      return
    }
   ...
}

This is something I've been noticing since Lottie 3.x.x where we have these kind of animations in our company app:
```
        let splashAnimation = AnimationView(name: "splash")
        splashAnimation?.contentMode = .scaleAspectFit
        view.addSubview(splashAnimation!)

         splashAnimation?.play(fromMarker: "1", toMarker: "2", loopMode: .playOnce, completion: { _ in
            self.splashAnimation?.play(fromMarker: "2", toMarker: "3", loopMode: .playOnce, completion: { _ in
                self.splashAnimation?.play(fromMarker: "2", toMarker: "3", loopMode: .playOnce, completion: { _ in
                    self.splashAnimation?.play(fromMarker: "3", toMarker: "4", loopMode: .playOnce, completion: { _ in
                        DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
                            self.splashAnimation?.play(fromMarker: "4", toMarker: "5", loopMode: .playOnce, completion: { _ in
                                DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
                                    self.route()
                                }
                            })
                        }
                    })
                })
            })
        })
```

**Comment by calda** [comment]
Can you share a sample project that reproduces this issue?

**Reply by oliveira-marcio** [reply]
Hey calda, thanks for answering. Actually, I only had this issue once when I tried to wrap up the snippet code above with checked continuations to make use of structured concurrency. Something like that:
```
extension AnimationView {
    func play(fromMarker: String? = nil,
              toMarker: String,
              loopMode: LottieLoopMode? = nil) async {
        await withCheckedContinuation { continuation in
            play(fromMarker: fromMarker,
                 toMarker: toMarker,
                 loopMode: loopMode) { _ in
                continuation.resume()
            }
        }
    }
}

...

await splashAnimation?.play(fromMarker: "1", toMarker: "2", loopMode: .playOnce)
await splashAnimation?.play(fromMarker: "2", toMarker: "3", loopMode: .playOnce)
await splashAnimation?.play(fromMarker: "2", toMarker: "3", loopMode: .playOnce)
await splashAnimation?.play(fromMarker: "3", toMarker: "4", loopMode: .playOnce)
try? await Task.sleep(nanoseconds: 1_000_000_000)
await splashAnimation?.play(fromMarker: "4", toMarker: "5", loopMode: .playOnce)
try? await Task.sleep(nanoseconds: 1_000_000_000)
route()
```
And one of my colleagues could notice once a `SWIFT TASK CONTINUATION MISUSE` warning in the Xcode which means that the continuation wasn't resumed (after some research about this warning), which lead me to think that one of the completions blocks in the original `play()` method wasn't called for some reason and my best guess was one of those guards lets on it that earlier return without calling the completion.

But like I've said this just happened once, but I'm worried if it's possible that the animation gets stuck and the `route()` method in my project might never be reached.
